client/knode/validator/validator.go

```diff
@@ -578,20 +578,27 @@ func (val *validator) vote(vote *types.Vote) {
 
 	err = val.votingSystem.Add(addressVote)
 	if err != nil {
-		log.Warn("Failed to add own vote to voting table", "err", err)
+		log.Error("Failed to add own vote to voting table",
+			"err", err, "blockHash", addressVote.Vote().BlockHash(), "hash", addressVote.Vote().Hash())
 	}
 }
 
 func (val *validator) AddBlockFragment(blockNumber *big.Int, round uint64, fragment *types.BlockFragment) error {
 	if !val.Validating() {
 		return ErrCantAddBlockFragmentNotValidating
 	}
-	val.blockFragments.Add(fragment)
+
+	if err := val.blockFragments.Add(fragment); err != nil {
+		err = errors.New("Failed to add a new block fragment: " + err.Error())
+		return err
+	}
 
 	if val.blockFragments.HasAll() {
 		block, err := val.blockFragments.Assemble()
 		if err != nil {
-			log.Crit("Failed to assemble the block", "err", err)
+			err = errors.New("Failed to assemble the block: " + err.Error())
```

Comment:
@acroca I'll keep `Error` here because `Assemble` doesn't change DB state and using `Crit` leads to the possibility to kill any node in the Network just sending it an incorrect block fragment.